src/meek.jl

```diff
@@ -0,0 +1,110 @@
+# http://proceedings.mlr.press/v89/katz19a/katz19a-supp.pdf
+# https://arxiv.org/pdf/1302.4972.pdf
+"""
+    meek_rules!(g; rule4=false)
+
+Apply Meek's rules 1-3 or 1-4 with rule4=true to orient edges in a 
+partially directed graph without creating cycles or new v-structures.
+Rule 4 is needed if edges are compelled/preoriented using external knowledge.
+"""
+function meek_rules!(g; rule4=false)
+    while true 
+        done = true
+        # Loop through all the edges in the graph (u-v)
+        for e in collect(edges(g)) # collect iterator as we are deleting
+            (u, v) = src(e), dst(e)
+            # We only need to update (still) undirected edges
+            !has_both(g, e) && continue
+
+            # check only case u->v, we'll check v->u later
+            if meek_rule1(g, u, v) || meek_rule2(g, u, v) || meek_rule3(g, u, v) || (rule4 && meek_rule4(g, u, v))
+                # Make u→v
+                remove!(g, v => u)
+                done = false
+            end
+        end
+        done && return g
+    end
+end
+
+"""
+    meek_rule1(dg, v, w)
+
+Rule 1: Orient v-w into v->w whenever there is u->v
+such that u and w are not adjacent
+(otherwise a new v-structure is created.)
+"""
+function meek_rule1(dg, v, w)
+    for u in inneighbors(dg, v)
+        has_edge(dg, v => u) && continue # not directed
+        isadjacent(dg, u, w) && continue
+        return true
+    end
+    return false
+end
+
+"""
+    meek_rule2(dg, v, w)
+
+Rule 2: Orient v-w into v->w whenever there is a chain v->k->w
+(otherwise a directed cycle is created.)
+"""
+function meek_rule2(dg, v, w)
+    outs = Int[]
+    for k in outneighbors(dg, v)
+        !has_edge(dg, k => v) && push!(outs, k)
+    end
+    ins = Int[]
+    for k in inneighbors(dg, w)
+        !has_edge(dg, w => k) && push!(ins, k)
+    end
+    if !disjoint_sorted(ins, outs)
+        return true
+    end
+    return false 
+end
+
+"""
+    meek_rule3(dg, v, w)
+
+Rule 3 (Diagonal): Orient v-w into v->w whenever there are two chains
+v-k->w and v-l->w such that k and l are nonadjacent
+(otherwise a new v-structure or a directed cycle is created.)
+"""
+function meek_rule3(dg, v, w)
+    fulls = [] # Find nodes k where v-k
+    for k in outneighbors(dg, v)
+        has_edge(dg, k => v) || continue 
+        # Skip if not k->w (or if not l->w)
+        if has_edge(dg, w => k) || !has_edge(dg, k => w)
+            continue
+        end
+        push!(fulls, k)
+    end
+    for (k, l) in combinations(fulls, 2) # FIXME: 
+        isadjacent(dg, k, l) && continue
+        return true
+    end
+    return false
+end
+
```

Comment:
Could this be written as:
```julia
function meek_rule3(dg, v, w)
        
    #given v-w, count the number of patterns that match v-x → w
    counter = 0
    for x in neighbors(g,v) #Loop though just the undirected adjacencies
        if isparent(dg,x,w)
            counter += 1
            #If we find two such paths, the Rule 3 pattern is matched
            if counter == 2
                return true
            end
        end
    end

    return false
end
```

Reply:
If you find v - x1 -> w and v - x2 -> w, it is still necessary to check that x1 and x2 are non-adjacent.